File: examples/example_ws_bidirection.py
import aiohttp
import asyncio
import json
import time
import hmac
import hashlib
import urllib.parse as parse
import uuid
import base64

# 请填写你的配置信息
SECRET_ID = "your_secret_id"
SECRET_KEY = "your_secret_key"
APP_ID = 0
SDK_APP_ID = 0

HOST = "flowtts.cloud.tencent.com"


def generate_signature(params):
    """生成签名"""
    sorted_params = sorted(params.items())
    sign_str = f"GET{HOST}/api/v1/flow_tts/bidirection?"
    sign_str += "&".join([f"{k}={v}" for k, v in sorted_params])
    
    signature = hmac.new(SECRET_KEY.encode('utf-8'), sign_str.encode('utf-8'), hashlib.sha1).digest()
    return base64.b64encode(signature).decode('utf-8')


def generate_url():
    """生成WebSocket连接URL"""
    connection_id = str(uuid.uuid4())
    timestamp = int(time.time())
    
    params = {
        "Action": "TextToSpeechBidirection",
        "AppId": APP_ID,
        "SecretId": SECRET_ID,
        "SdkAppId": SDK_APP_ID,
        "Timestamp": timestamp,
        "Expired": timestamp + 86400,
        "ConnectionId": connection_id,
    }
    
    params["Signature"] = generate_signature(params)
    query_string = "&".join([f"{k}={parse.quote(str(v))}" for k, v in sorted(params.items())])
    url = f"wss://{HOST}/api/v1/flow_tts/bidirection?{query_string}"
    
    return url, connection_id


class TTSWebSocketClient:
    def __init__(self):
        self.ws = None
        self.connection_id = None
        self.session_id = None
        self.session = None

    async def connect(self):
        """建立WebSocket连接"""
        url, self.connection_id = generate_url()
        print(f"连接URL: {url}")

        self.session = aiohttp.ClientSession()

        try:
            async with self.session.ws_connect(url) as ws:
                self.ws = ws
                print("WebSocket连接已建立")
                await self.start_session()
                await self.receive_messages()
        except Exception as e:
            print(f"连接错误: {e}")
        finally:
            await self.session.close()

    async def receive_messages(self):
        """接收WebSocket消息"""
        async for msg in self.ws:
            if msg.type == aiohttp.WSMsgType.TEXT:
                await self.handle_message(msg.data)
            elif msg.type == aiohttp.WSMsgType.ERROR:
                print(f"WebSocket错误: {self.ws.exception()}")
                break
            elif msg.type == aiohttp.WSMsgType.CLOSED:
                print("WebSocket连接已关闭")
                break

    async def handle_message(self, message):
        """处理接收到的消息"""
        msg = json.loads(message)
        event = msg.get("Event")
        print(f"\n收到事件: {event}")

        if event == "SessionStart":
            self.session_id = msg.get("SessionId")
            print(f"会话已开始，SessionId: {self.session_id}")
            asyncio.create_task(self.send_text_stream())

        elif event == "SentenceAudio":
            data = msg.get("Data", {})
            print(f"收到句子: {data.get('Sentence')} (音频: {len(data.get('Audio', ''))} 字符)")

        elif event == "SessionEnd":
            data = msg.get("Data", {})
            print(f"会话结束 - 句子数: {data.get('TotalSentences')}, 时长: {data.get('TotalDuration')}秒")
            await self.ws.close()

        elif event == "SessionError":
            error_data = msg.get("Data", {})
            print(f"会话错误: {error_data.get('ErrorCode')} - {error_data.get('ErrorMessage')}")

        elif event == "SentenceError":
            error_data = msg.get("Data", {})
            print(f"句子错误: {error_data}")

    async def start_session(self):
        """开始会话"""
        message = {
            "Event": "StartSession",
            "ConnectionId": self.connection_id,
            "SessionId": "",
            "MessageId": str(uuid.uuid4()),
            "Data": {
                "Voice": {
                    "VoiceId": "v-male-s5NqE0rZ"
                }
            }
        }

        await self.ws.send_str(json.dumps(message, ensure_ascii=False))
        print("已发送StartSession")

    async def send_text_stream(self):
        """流式发送文本"""
        texts = ["今天天气", "真好！", "你那边", "怎么样？", "我这边阳光明媚。"]

        for i, text in enumerate(texts):
            await asyncio.sleep(1)
            message = {
                "Event": "ContinueSession",
                "ConnectionId": self.connection_id,
                "SessionId": self.session_id,
                "MessageId": str(uuid.uuid4()),
                "Data": {"Text": text}
            }
            await self.ws.send_str(json.dumps(message, ensure_ascii=False))
            print(f"已发送文本 [{i+1}/{len(texts)}]: {text}")

        await asyncio.sleep(1)
        await self.finish_session()

    async def finish_session(self):
        """结束会话"""
        message = {
            "Event": "FinishSession",
            "ConnectionId": self.connection_id,
            "SessionId": self.session_id,
            "MessageId": str(uuid.uuid4()),
            "Data": {}
        }
        await self.ws.send_str(json.dumps(message, ensure_ascii=False))
        print("已发送FinishSession")


async def main():
    client = TTSWebSocketClient()
    await client.connect()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/nodejs/example_ws_bidirection.js
#!/usr/bin/env node
/**
 * FlowTTS WebSocket Bidirectional Streaming Example
 *
 * Corresponds to examples/python/example_ws_bidirection.py
 *
 * This example does NOT use the Tencent Cloud SDK.
 * It implements the WebSocket protocol directly using the "ws" library.
 */

import crypto from "crypto";
import WebSocket from "ws";
import { loadConfig } from "./utils.js";

const HOST = "flowtts.cloud.tencent.com";

function generateSignature(params, secretKey) {
  const sortedParams = Object.entries(params).sort(([a], [b]) =>
    a.localeCompare(b)
  );
  let signStr = `GET${HOST}/api/v1/flow_tts/bidirection?`;
  signStr += sortedParams.map(([k, v]) => `${k}=${v}`).join("&");

  const hmac = crypto.createHmac("sha1", secretKey);
  hmac.update(signStr);
  return hmac.digest("base64");
}

function generateUrl(cfg) {
  const connectionId = crypto.randomUUID();
  const timestamp = Math.floor(Date.now() / 1000);

  const params = {
    Action: "TextToSpeechBidirection",
    AppId: cfg.sdkAppId,
    SecretId: cfg.secretId,
    SdkAppId: cfg.sdkAppId,
    Timestamp: timestamp,
    Expired: timestamp + 86400,
    ConnectionId: connectionId,
  };

  params.Signature = generateSignature(params, cfg.secretKey);

  const queryString = Object.entries(params)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([k, v]) => `${k}=${encodeURIComponent(String(v))}`)
    .join("&");

  const url = `wss://${HOST}/api/v1/flow_tts/bidirection?${queryString}`;
  return { url, connectionId };
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

class TTSWebSocketClient {
  constructor() {
    this.ws = null;
    this.connectionId = null;
    this.sessionId = null;
  }

  connect(cfg) {
    return new Promise((resolve, reject) => {
      const { url, connectionId } = generateUrl(cfg);
      this.connectionId = connectionId;
      console.log(`连接URL: ${url}`);

      this.ws = new WebSocket(url);

      this.ws.on("open", async () => {
        console.log("WebSocket连接已建立");
        await this.startSession();
      });

      this.ws.on("message", async (data) => {
        await this.handleMessage(data.toString());
      });

      this.ws.on("error", (err) => {
        console.log(`WebSocket错误: ${err.message}`);
        reject(err);
      });

      this.ws.on("close", () => {
        console.log("WebSocket连接已关闭");
        resolve();
      });
    });
  }

  async handleMessage(message) {
    const msg = JSON.parse(message);
    const event = msg.Event;
    console.log(`\n收到事件: ${event}`);

    if (event === "SessionStart") {
      this.sessionId = msg.SessionId;
      console.log(`会话已开始，SessionId: ${this.sessionId}`);
      // Start sending text stream (non-blocking)
      this.sendTextStream().catch(console.error);
    } else if (event === "SentenceAudio") {
      const data = msg.Data || {};
      const audioLen = (data.Audio || "").length;
      console.log(`收到句子: ${data.Sentence} (音频: ${audioLen} 字符)`);
    } else if (event === "SessionEnd") {
      const data = msg.Data || {};
      console.log(
        `会话结束 - 句子数: ${data.TotalSentences}, 时长: ${data.TotalDuration}秒`
      );
      this.ws.close();
    } else if (event === "SessionError") {
      const data = msg.Data || {};
      console.log(`会话错误: ${data.ErrorCode} - ${data.ErrorMessage}`);
    } else if (event === "SentenceError") {
      const data = msg.Data || {};
      console.log(`句子错误: ${JSON.stringify(data)}`);
    }
  }

  async startSession() {
    const message = {
      Event: "StartSession",
      ConnectionId: this.connectionId,
      SessionId: "",
      MessageId: crypto.randomUUID(),
      Data: {
        Voice: {
          VoiceId: "v-male-s5NqE0rZ",
        },
      },
    };

    this.ws.send(JSON.stringify(message));
    console.log("已发送StartSession");
  }

  async sendTextStream() {
    const texts = [
      "今天天气",
      "真好！",
      "你那边",
      "怎么样？",
      "我这边阳光明媚。",
    ];

    for (let i = 0; i < texts.length; i++) {
      await sleep(1000);
      const message = {
        Event: "ContinueSession",
        ConnectionId: this.connectionId,
        SessionId: this.sessionId,
        MessageId: crypto.randomUUID(),
        Data: { Text: texts[i] },
      };
      this.ws.send(JSON.stringify(message));
      console.log(`已发送文本 [${i + 1}/${texts.length}]: ${texts[i]}`);
    }

    await sleep(1000);
    await this.finishSession();
  }

  async finishSession() {
    const message = {
      Event: "FinishSession",
      ConnectionId: this.connectionId,
      SessionId: this.sessionId,
      MessageId: crypto.randomUUID(),
      Data: {},
    };
    this.ws.send(JSON.stringify(message));
    console.log("已发送FinishSession");
  }
}

async function main() {
  const cfg = loadConfig();
  const client = new TTSWebSocketClient();
  await client.connect(cfg);
}

main().catch(console.error);
